data/slash/multipleHeaderSigningProof.go

```diff
@@ -75,36 +98,83 @@ func NewMultipleSigningProof(slashResult map[string]SlashingResult) (MultipleSig
 		return nil, data.ErrNilSlashResult
 	}
 
-	pubKeys := make([][]byte, 0, len(slashResult))
-	levels := make(map[string]ThreatLevel, len(slashResult))
-	headers := make(map[string]HeadersV2, len(slashResult))
+	headersInfo, err := getAllUniqueHeaders(slashResult)
+	if err != nil {
+		return nil, err
+	}
+	sortedHeaders, err := sortAndGetHeadersV2(headersInfo)
+	if err != nil {
+		return nil, err
+	}
 
-	sortedPubKeys := getSortedPubKeys(slashResult)
-	for _, pubKey := range sortedPubKeys {
-		pubKeys = append(pubKeys, []byte(pubKey))
-		levels[pubKey] = slashResult[pubKey].SlashingLevel
+	hashIndexMap := calcHashIndexMap(headersInfo)
+	signersSlashData := computeSignersSlashData(hashIndexMap, slashResult)
 
-		sortedHeaders, err := getSortedHeadersV2(slashResult[pubKey].Headers)
-		if err != nil {
-			return nil, err
+	return &MultipleHeaderSigningProof{
+		HeadersV2:        *sortedHeaders,
+		SignersSlashData: signersSlashData,
+	}, nil
+}
+
+func getAllUniqueHeaders(slashResult map[string]SlashingResult) ([]data.HeaderInfoHandler, error) {
+	headersInfo := make([]data.HeaderInfoHandler, 0, len(slashResult))
+	hashes := make(map[string]struct{})
+
+	for pubKey, res := range slashResult {
+		hashesPerPubKey := make(map[string]struct{})
+		for _, currHeaderInfo := range res.Headers {
+			if currHeaderInfo == nil {
```

Comment:
I have seen this same logic to validate a headerInfo in the sorting function in common.go
I suggest to extract this into a validation function with parameters in common, and reuse in both places (e.g from L126->L140)

Reply:
Good suggestion. Done